File: c++/isBalanced.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <cstdlib>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

int depth = 0;
int dep[1000];
int cnt = 0;

int getDepth(TreeNode* root, bool &result) {
    if (!root) return 0;
    if (!result) return 0;

    int ldep = getDepth(root->left, result);
    int rdep = getDepth(root->right, result);
    if (!result) return 0;

    result = abs(ldep - rdep) <= 1;
    return max(ldep, rdep) + 1;
}

bool isBalanced(TreeNode *root) {
    bool result = true;
    getDepth(root, result);
    return result;
}

void preOrder(TreeNode* root) 
{
    if (!root) 
    {
        cout << "#\t" ;
        return;
    }

    cout << root->val << "\t";
    preOrder(root->left);
    preOrder(root->right);
}

int main() {
    int a[4] = {-1, 2, -1, 3};
    TreeNode* root = new TreeNode(1);
    TreeNode* tmp, *curr;
    curr = root;
        
    for (int i = 0; i < 4; ++i){
        if (!(i >> 1) && a[i] != -1) 
        { 
            tmp = new TreeNode(a[i]);
            curr->left = tmp;
            curr = curr->left;
        } else if (!(i >> 1)) {
                curr->left = NULL;
        } else if (a[i] != -1) {            
            tmp = new TreeNode(a[i]);
            curr->right = tmp;
            curr = curr->right;
        } else {
            curr->right = NULL;
        }
    }
    preOrder(root);
    cout << isBalanced(root) << endl;
    
    return 0;
}

File: java/isBalanced.java
import com.TreeNode;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class isBalanced {
    public static class Solution {

        private void helper(TreeNode root, List<Integer> res, int idx) {
            int cnt = res.get(0);
            if (root == null) return;
            else if (root.left == null && root.right == null) {
                if (idx > cnt)
                    res.add(0, idx);
                return;
            }
            helper(root.left, res, idx + 1);
            helper(root.right, res, idx + 1);
        }
        
        private int maxDepth(TreeNode root) {
            if (root == null) return 0;
            else if (root.left == null && root.right == null) return 1;

            List<Integer> res = new ArrayList<Integer>();
            int cnt = 1;
            res.add(cnt);
            helper(root, res, 1);
            return res.get(0);
        }

        public boolean isBalanced(TreeNode root) {
            if (root == null) return true;
            else if (root.left == null && root.right == null) return true;
            else if (root.left == null)
                return root.right.left == null && root.right.right == null;
            else if (root.right == null)
                return root.left.left == null && root.left.right == null;
            
            if (!isBalanced(root.left) || !isBalanced(root.right)) return false;
            // must make sure they differ at most by one
            return Math.abs(maxDepth(root.left) - maxDepth(root.right)) <= 1;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {1, 2, 2, 3, 3, -1, -1, 4, 4};
        TreeNode root = new TreeNode(a[0]);
        root.buildTree(root, a);
        root.levelPrintTree(root);
        boolean res = result.isBalanced(root);

        System.out.println(res);
    }
}
